### ISSUE TITLE
priority queues?

### ISSUE THREAD

**Issue by tobowers** [asynkron/protoactor-go#379]
What's the best way to implement a priority queue using protoactor? I was looking at creating a new mailbox type, but was unsure exactly how to implement that (since defaultMailbox is not exported).

I was then trying to think through how to create a set of different actors, but that seems hard too :).

**Comment by dennis-tra** [comment]
Hi @tobowers,

while there is a `WithMailbox` method on `*actor.Props` I think this is only meant to choose between the already implemented mailboxes

* unbounded
* bounded
* bounded dropping

and not to provide your own one. (maintainers, please correct me on this :+1:)

The logic of defaultMailbox seems too intricate to reimplement it outside this library. So I think the only reasonable way would be to jump into the `mailbox` package and implement your own priority mailbox there. Akka seems to have the notion of an unbounded priority mailbox already [1] - maybe this could give valuable insights.

You may already have your own thoughts but this is what goes through my mind atm:

At a first glance I would think of something like an `unbounded_priority.go` in the `mailbox` package:
```go
package mailbox

import (
	// implement 
)

type unboundedPriorityMailboxQueue struct {
	// implement 
}

func (q *unboundedPriorityMailboxQueue) Push(m interface{}) {
	// implement 
}

func (q *unboundedPriorityMailboxQueue) Pop() interface{} {
	// implement 
}

// Unbounded returns a producer which creates an unbounded mailbox
func UnboundedPriority(mailboxStats ...Statistics) Producer {
	return func() Mailbox {
		q := &unboundedPriorityMailboxQueue{
			// implement
		}
		return &defaultMailbox{
			systemMailbox: mpsc.New(),
			userMailbox:   q,
			mailboxStats:  mailboxStats,
		}
	}
}
```

One would probably need an interface that the messages should conform to which has one method that should return the priority of a given message.

[1]: https://doc.akka.io/japi/akka/current/akka/dispatch/UnboundedPriorityMailbox.html

**Comment by tobowers** [comment]
Thanks for the thoughtful comment! I think this is difficult to implement in a generic way because specifying the priority in the message can be tricky. It's interesting because the mailbox *is* a priority mailbox already (system messages come first). In logic it would probably look something like that but with more "levels" (system always comes first, then this level then the next).

However the logic of "where do these messages go" is tough in Go without introspection.

**Comment by dennis-tra** [comment]
I have implemented something similar to your concept of "levels" in PR #350 [remote/endpoint_writer_mailbox.go:L25](https://github.com/AsynkronIT/protoactor-go/pull/350/files#diff-66aa10f50b416dccd0536754475dbb1f) for the `endpointWriterMailbox`. The relevant excerpt of the PR-description:
> I added a tiny (100 entries) priority mailbox which gets drained before the user message mailbox and can only contain *actor.ReceiveTimeout and *EndpointTerminated messages

This is probably something similar to what you want - just a mailbox that gets drained before another mailbox.

However, you're right that it is tricky to implement this concept in a generic way. Specifying the priority on messages could be done by also defining an interface which the messages must conform to. With my last sentence above I meant something like this:
```go
type PriorityMessage interface {
	GetPriority() int32
}
```

and 

```go
func (q *unboundedPriorityMailboxQueue) Push(m interface{}) {
	// type cast to PriorityMessage
        // if successful -> grab priority and put it into the queue accordingly
        // if unsuccessful -> use a default priority
}
```

```go
type MyImportantMessage struct {
	payload string
}

func (msg *MyImportantMessage) GetPriority() int32 {
	return 1000 // hard coded or based on payload
}
```

**Comment by tobowers** [comment]
yeah that looks good the question of "what do I do with 3 messages of priority 1, 50, 1000" is also a bit difficult :). But yeah your tiny priority mailbox is exactly what I want in my specific use case.

I have a stream of messages coming in and sometimes I need to do a state change before processing the rest of them... I don't want to queue that state change up at the end of them.

**Comment by rogeralsing** [comment]
Closing due to inactivity. Reopen if still relevant.
